core/error/error_macros.h

```diff
@@ -786,8 +786,18 @@ void _err_flush_stdout();
 		((void)0)
 
 /**
- * This should be a 'free' assert for program flow and should not be needed in any releases,
- *  only used in dev builds.
+ * Note: IN MOST CASES YOU SHOULD NOT USE THIS MACRO.
+ * Do not use unless you understand the trade-offs.
+ *
+ * DEV macros will be compiled out in releases, they are wrapped in DEV_ENABLED.
+ *
+ * Prefer WARNINGS / ERR_FAIL macros (which fail without crashing) - ERR_FAIL should be used in most cases.
+ * Then CRASH_NOW_MSG macros (on rare occasions where error cannot be recovered).
+ *
+ * DEV_ASSERT should generally only be used in (usually both these conditions should be met):
+ * 1) Bottleneck code where a check in release would be too expensive
+ * 2) Situations where the check would fail obviously and straight away
+ * (to prevent merging code that would trigger DEV_ASSERTs, which would annoy other contributors).
```

Comment:
Just expanding a bit:

```suggestion
 * 1) Bottleneck code where a check in release would be too expensive
 * 2) Situations where the check would fail obviously and straight away
 *    during the mainteanance of the code, like strict pre/post-conditions,
 *    that assist the code writer and give clues to the code reader
 * (to prevent merging code that would trigger DEV_ASSERTs, which would annoy other contributors).
```

(Or a better worded version of that.)

Reply:
I'm slightly unsure on this wording. :thinking: 

One big danger is ending up with contributors using `DEV_ASSERT` for the pre-post conditions where they _should_ be using `ERR_FAIL` / `WARN_PRINT`.

Although your added wording is true, it has not been similarly written for the `ERR_FAIL` macros (and would perhaps be needed to be repeated in too many places for this to be practical), so again could lead contributors again to wrong usage (they might say that this macro explicitly mentions pre/post conditions, whereas `ERR_FAIL` does not...).

Reply:
```
 * 2) Situations where the check would fail obviously and straight away
 *    during the mainteanance of the code (i.e. strict conditions that should
 *    be true no matter what) and that can't fail for other contributors
 *    once the code is finished and merged.
 ```
❔

Reply:
Have changed to this. :+1: